From 393da557977372694e5c73f2f5429721edb962d7 Mon Sep 17 00:00:00 2001
From: Pauli Virtanen <pav@iki.fi>
Date: Thu, 25 Jun 2026 17:35:35 +0300
Subject: [PATCH] bluez5: backend-native: fix multiple profile registration

Handling pending RegisterProfile callbacks was wrong as it forgot that
there can be multiple profiles to be registered.

Fix the handling by allowing several concurrent register callbacks.

diff --git a/spa/plugins/bluez5/backend-native.c b/spa/plugins/bluez5/backend-native.c
index 1cb2235..ece7571 100644
--- a/spa/plugins/bluez5/backend-native.c
+++ b/spa/plugins/bluez5/backend-native.c
@@ -106,6 +106,8 @@ struct impl {
 	struct spa_dbus *dbus;
 	DBusConnection *conn;
 
+	DBusPendingCall *pending_register_profile[4];
+
 	const struct media_codec * const * codecs;
 
 #define DEFAULT_ENABLED_PROFILES (SPA_BT_PROFILE_HFP_HF | SPA_BT_PROFILE_HFP_AG)
@@ -3669,11 +3671,40 @@ static DBusHandlerResult profile_handler(DBusConnection *c, DBusMessage *m, void
 	return res;
 }
 
+static bool pending_register_profile_push(struct impl *backend, DBusPendingCall *pending)
+{
+	SPA_FOR_EACH_ELEMENT_VAR(backend->pending_register_profile, p) {
+		if (!*p) {
+			*p = pending;
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static DBusMessage *pending_register_profile_pop(struct impl *backend, DBusPendingCall *pending)
+{
+	SPA_FOR_EACH_ELEMENT_VAR(backend->pending_register_profile, p) {
+		if (*p == pending)
+			return steal_reply_and_unref(p);
+	}
+
+	spa_assert_not_reached();
+	return NULL;
+}
+
+static void pending_register_profile_clear(struct impl *backend)
+{
+	SPA_FOR_EACH_ELEMENT_VAR(backend->pending_register_profile, p)
+		cancel_and_unref(p);
+}
+
 static void register_profile_reply(DBusPendingCall *pending, void *user_data)
 {
 	struct impl *backend = user_data;
 
-	spa_autoptr(DBusMessage) r = steal_reply_and_unref(&pending);
+	spa_autoptr(DBusMessage) r = pending_register_profile_pop(backend, pending);
 	if (r == NULL)
 		return;
 
@@ -3690,6 +3721,8 @@ static void register_profile_reply(DBusPendingCall *pending, void *user_data)
 				dbus_message_get_error_name(r));
 		return;
 	}
+
+	spa_log_debug(backend->log, "RegisterProfile() complete");
 }
 
 static int register_profile(struct impl *backend, const char *profile, const char *uuid)
@@ -3698,6 +3731,7 @@ static int register_profile(struct impl *backend, const char *profile, const cha
 	DBusMessageIter it[4];
 	dbus_bool_t autoconnect;
 	dbus_uint16_t version, chan, features;
+	DBusPendingCall *pending;
 	const char *str;
 
 	if (!(backend->enabled_profiles & spa_bt_profile_from_uuid(uuid)))
@@ -3801,8 +3835,12 @@ static int register_profile(struct impl *backend, const char *profile, const cha
 	}
 	dbus_message_iter_close_container(&it[0], &it[1]);
 
-	if (!send_with_reply(backend->conn, m, register_profile_reply, backend))
+	pending = send_with_reply(backend->conn, m, register_profile_reply, backend);
+	if (!pending || !pending_register_profile_push(backend, pending)) {
+		spa_log_error(backend->log, "Failed to start RegisterProfile() %s", profile);
+		cancel_and_unref(&pending);
 		return -EIO;
+	}
 
 	return 0;
 }
@@ -3868,6 +3906,8 @@ static int backend_native_unregister_profiles(void *data)
 {
 	struct impl *backend = data;
 
+	pending_register_profile_clear(backend);
+
 	sco_close(backend);
 
 #ifdef HAVE_BLUEZ_5_BACKEND_HSP_NATIVE
@@ -4048,6 +4088,8 @@ static int backend_native_free(void *data)
 
 	struct rfcomm *rfcomm;
 
+	pending_register_profile_clear(backend);
+
 	sco_close(backend);
 
 	if (backend->modemmanager) {
